sql/expression/function/aggregation/window_block.go

```diff
@@ -0,0 +1,319 @@
+	return Aggregation{fn: a, framer: f}
+}
+
+type windowBlockIter struct {
+	ctx   *sql.Context
+	pos   int
+	child sql.RowIter
+
+	partitionBy []sql.Expression
+	sortBy      sql.SortFields
+	aggs        []Aggregation
+
+	partitions       []sql.WindowInterval
+	currentPartition sql.WindowInterval
+	partitionIdx     int
+	done             bool
+
+	input, output     sql.WindowBuffer
+	outputOrdering    []int
+	outputIdx         []int
+	outputOrderingPos int
+}
+
+var _ sql.RowIter = (*windowBlockIter)(nil)
+var _ sql.Disposable = (*windowBlockIter)(nil)
+
+func NewWindowBlockIter(partitionBy []sql.Expression, sortBy sql.SortFields, aggs []Aggregation, child sql.RowIter) *windowBlockIter {
+	return &windowBlockIter{
+		partitionBy:  partitionBy,
+		sortBy:       sortBy,
+		aggs:         aggs,
+		child:        child,
+		partitionIdx: -1,
+	}
+}
+
+func (i *windowBlockIter) Close(ctx *sql.Context) error {
+	i.Dispose()
+	i.input = nil
+	return i.child.Close(ctx)
+}
+
+func (i *windowBlockIter) Dispose() {
+	for _, a := range i.aggs {
+		a.fn.Dispose()
+	}
+}
+
+func (i *windowBlockIter) Next(ctx *sql.Context) (sql.Row, error) {
+	var err error
+	if i.output == nil {
+		i.input, i.outputOrdering, err = i.materializeInput(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		i.partitions, err = i.initializePartitions(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		i.output, err = i.materializeOutput(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		err = i.sortAndFilterOutput()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if i.pos > len(i.output)-1 {
+		return nil, io.EOF
+	}
+
+	defer func() { i.pos++ }()
+
+	return i.output[i.pos], nil
+}
+
+func (i *windowBlockIter) materializeOutput(ctx *sql.Context) (sql.WindowBuffer, error) {
+	// handle nil input specially if no partition clause
+	// ex: COUNT(*) on nil rows returns 0, not nil
+	if len(i.input) == 0 && len(i.partitionBy) > 0 {
+		return nil, io.EOF
+	}
+
+	output := make(sql.WindowBuffer, 0, len(i.input))
+	var row sql.Row
+	var err error
+	for {
+		row, err = i.compute(ctx)
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		output = append(output, row)
+	}
+
+	return output, nil
+}
+
+func (i *windowBlockIter) sortAndFilterOutput() error {
+	// TODO: move sort by above aggregations
+	// we could cycle sort this for windows (not group by, unless number
+	// of group by partitions = number of rows)
+	if len(i.output) == 0 {
+		return nil
+	}
+
+	originalOrderIdx := len(i.output[0]) - 1
+	sort.SliceStable(i.output, func(j, k int) bool {
+		return i.output[j][originalOrderIdx].(int) < i.output[k][originalOrderIdx].(int)
+	})
+
+	for j, row := range i.output {
+		i.output[j] = row[:originalOrderIdx]
+	}
+
+	return nil
+}
+
+func (i *windowBlockIter) compute(ctx *sql.Context) (sql.Row, error) {
+	var row = make(sql.Row, len(i.aggs)+1)
+
+	for j, agg := range i.aggs {
+		interval, err := agg.framer.Next()
```

Comment:
This looks wrong (and if it's correct it's very misleading): it looks like you're advancing to the next frame for each aggregate function, rather than once per output row. Should be tests that would catch this.

Probably want to advance the frame strictly outside this method.

Reply:
every aggregation has its own frame generator. i think the difference is that a set of aggregations' `partitions` are managed globally in a block, but within the block each aggregation can be framed differently as soon as i turn on framing.